src/enclave/enclave_oe.c

```diff
@@ -1,27 +1,144 @@
+
+    if (sgxlkl_enclave_state.shared_memory.env && cfg->num_host_import_env > 0)
+    {
+        state->imported_env = malloc(sizeof(char*) * cfg->num_host_import_env);
+        if (!state->imported_env)
+            sgxlkl_fail("out of memory\n");
+
+        for (size_t i = 0; i < cfg->num_host_import_env; i++)
+        {
+            const char* name = cfg->host_import_env[i];
+            for (char* const* p = sgxlkl_enclave_state.shared_memory.env;
+                 p && *p != NULL;
+                 p++)
+            {
+                size_t n = strlen(name);
+                if (strncmp(name, *p, n) == 0 && (*p)[n] == '=')
+                {
+                    const char* str = *p;
+                    size_t len = strlen(str);
+                    char* cpy = malloc(len + 1);
+                    if (!cpy)
+                        sgxlkl_fail("out of memory\n");
+                    memcpy(cpy, str, len + 1);
+                    state->imported_env[state->num_imported_env++] = cpy;
+                }
+            }
+        }
+    }
+
+    size_t total_size = 0;
+    size_t total_count = 1;
+    for (size_t i = 0; i < cfg->num_args; i++)
+        total_size += strlen(cfg->args[i]) + 1;
+    total_count += cfg->num_args + 1;
+    for (size_t i = 0; i < cfg->num_env; i++)
+        total_size += strlen(cfg->env[i]) + 1;
+    total_count += cfg->num_env + 1;
+    for (size_t i = 0; i < state->num_imported_env; i++)
+        total_size += strlen(state->imported_env[i]) + 1;
+    total_count += state->num_imported_env + 1;
+    total_count += 1; // auxv terminator
+    total_count += 1; // platform-independent stuff terminator
+
+    char* buf = calloc(total_size, sizeof(char));
+    char** out = calloc(total_count, sizeof(char*));
+
+    size_t j = 0;
+    char* buf_ptr = buf;
+
+#define ADD_STRING(S)               \
+    {                               \
+        size_t len = strlen(S) + 1; \
+        memcpy(buf_ptr, (S), len);  \
+        out[j++] = buf_ptr;         \
+        buf_ptr += len;             \
+    }
+
+    elf64_stack_t* stack = &sgxlkl_enclave_state.elf64_stack;
 
-extern _Atomic(enum sgxlkl_libc_state) __libc_state;
+    // argv
+    stack->argv = out;
+    for (size_t i = 0; i < cfg->num_args; i++)
+        ADD_STRING(cfg->args[i]);
+    stack->argc = j;
+    out[j++] = NULL;
 
-sgxlkl_config_t* sgxlkl_enclave;
+    // envp
+    stack->envp = out + j;
+    for (size_t i = 0; i < cfg->num_env; i++)
+        ADD_STRING(cfg->env[i]);
+    for (size_t i = 0; i < state->num_imported_env; i++)
+        // Is this the right order for imported vars?
+        ADD_STRING(state->imported_env[i]);
+    out[j++] = NULL;
+
+    // auxv
+    stack->auxv = (Elf64_auxv_t**)(out + j);
+    for (size_t i = 0; i < cfg->num_auxv; i++)
+    {
+        out[j++] = (char*)cfg->auxv[i].a_type;
+        out[j++] = (char*)cfg->auxv[i].a_un.a_val;
+    }
+    out[j++] = NULL;
+
+    // TODO: platform independent things?
+    out[j++] = NULL;
```

Comment:
All of the strings in argv, environ, and the ELF aux args vector go here.  See Figure 3.11 of https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf

Please coordinate with @vtikoo, who is having to initialise the first userspace thread with this layout as part of the pthread / lthread work.

Reply:
I'm not going to dive into that if @vtikoo is already working on it. I'm happy for this to get replaced completely, this was just a way for me to translate from config settings to a stack that works for our tests.